iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,261 @@
+/// 
+/// Property under the root component:
+/// "reported": {
+///     "targetTemperature": {
+///         "value": 21.3,
+///         "ac": 200,
+///         "av": 3,
+///         "ad": "complete"
+///     }
+/// }
+/// 
+/// Property under a specified component:
+/// "reported": {
+///   "thermostat1": {
+///     "__t": "c",
+///     "targetTemperature": {
+///       "value": 23,
+///       "ac": 200,
+///       "av": 3,
+///       "ad": "complete"
+///     }
+///   }
+/// }
+/// </summary>
+/// <param name="propertyName">The name of the twin property to update.</param>
+/// <param name="propertyValue">The value of the twin property to update.</param>
+/// <param name="componentName">The name of the component in which the property is defined. Can be null for property defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdatePropertyAsync(string propertyName, WritableProperty propertyValue, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Subscribe to writable property event for all properties under any component.
+///
+/// Payload received for the root component:
+/// "desired" :
+/// {
+///   "targetTemperature" : 21.3,
+///   "targetHumidity" : 80
+/// },
+/// "$version" : 3
+///
+/// Payload received for a specific component:
+/// "desired": {
+///   "thermostat1": {
+///     "__t": "c",
+///     "targetTemperature": 21.3,
+///     "targetHumidity": 80
+///   }
+/// },
+/// "$version" : 3
+/// 
+/// </summary>
+/// <param name="propertyActionAsTwinCollection">The action to be taken when a writeable property event is received.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public void SubscribeToWritablePropertyEvent(Action<TwinCollection> propertyActionAsTwinCollection, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Retrieve the device properties.
+/// </summary>
+/// <returns>The device properties.</returns>
+public Task<Twin> GetPropertiesAsync(CancellationToken cancellationToken = default);
+```
+
+<details>
+<summary>WritableProperty</summary>
+
+```csharp
+/// <summary>
+/// Empty constructor.
+/// </summary>
+public WritablePropertyResponse() { }
+
+/// <summary>
+/// Convenience constructor for specifying the properties.
+/// </summary>
+/// <param name="propertyValue">The unserialized property value.</param>
+/// <param name="ackCode">The acknowledgement code, usually an HTTP Status Code e.g. 200, 400.</param>
+/// <param name="ackVersion">The acknowledgement version, as supplied in the property update request.</param>
+/// <param name="ackDescription">The acknowledgement description, an optional, human-readable message about the result of the property update.</param>
+public WritablePropertyResponse(object propertyValue, int ackCode, long ackVersion, string ackDescription = null)
```

Comment:
> Why object here for the param, but dynamic in the property?
Why this over generic types?

object vs dynamic vs T

Reply:
I will vote for T